core/src/main/scala/za/co/absa/spline/harvester/plugin/embedded/SnowflakePlugin.scala

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright 2021 ABSA Group Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.spline.harvester.plugin.embedded
+
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.execution.datasources.SaveIntoDataSourceCommand
+import za.co.absa.spline.commons.reflect.extractors.SafeTypeMatchingExtractor
+import za.co.absa.spline.harvester.builder.SourceIdentifier
+import za.co.absa.spline.harvester.plugin.Plugin.{Precedence, WriteNodeInfo}
+import za.co.absa.spline.harvester.plugin.embedded.SnowflakePlugin._
+import za.co.absa.spline.harvester.plugin.{Plugin, RelationProviderProcessing}
+
+import javax.annotation.Priority
+import scala.language.reflectiveCalls
+
+
+@Priority(Precedence.Normal)
+class SnowflakePlugin(spark: SparkSession)
+  extends Plugin
+    with RelationProviderProcessing {
+
+  import za.co.absa.spline.commons.ExtractorImplicits._
+
+  override def relationProviderProcessor: PartialFunction[(AnyRef, SaveIntoDataSourceCommand), WriteNodeInfo] = {
+    case (rp, cmd) if rp == "net.snowflake.spark.snowflake.DefaultSource" || SnowflakeSourceExtractor.matches(rp) =>
+      val url: String = cmd.options("sfUrl")
+      val warehouse: String = cmd.options("sfWarehouse")
+      val database: String = cmd.options("sfDatabase")
+      val schema: String = cmd.options("sfSchema")
+      val table: String = cmd.options("dbtable")
+
+      WriteNodeInfo(asSourceId(url, warehouse, database, schema, table), cmd.mode, cmd.query, cmd.options)
```

Comment:
This captures a WRITE operation, but what about a READ one?
Is `DataSourceV2` supported by the Snowflake connector?

Reply:
I don't know have an explicit answer to your question. However, I did some digging on capturing READ, and I wasn't able to find a way to capture the data source URI given the underlying `net.snowflake.spark.snowflake.SnowflakeRelation` that gets matched on when the logical plan includes reading from a snowflake source. 

This is what was available in that `SnowflakeRelation` object on a READ test I executed:

```
{
  "schema": {
    "fields": [
      {
        "name": "WORKORDERID",
        "type": "StringType",
        "nullable": true
      },
      {
        "name": "PRODUCTID",
        "type": "StringType",
        "nullable": true
      },
      {
        "name": "ORDERQTY",
        "type": "StringType",
        "nullable": true
      },
      {
        "name": "SCRAPPEDQTY",
        "type": "StringType",
        "nullable": true
      },
      {
        "name": "STARTDATE",
        "type": "StringType",
        "nullable": true
      },
      {
        "name": "ENDDATE",
        "type": "StringType",
        "nullable": true
      },
      {
        "name": "MODIFIEDDATE",
        "type": "StringType",
        "nullable": true
      },
      {
        "name": "BONUS_PERCENT",
        "type": "DoubleType",
        "nullable": true
      }
    ]
  },
  "scanMetrics": {
    "numBytesRead": {
      "id": 1,
      "name": "number of bytes read",
      "value": 7197975
    },
    "numRecordsRead": {
      "id": 2,
      "name": "number of records read",
      "value": 72591
    }
  },
  "writeMetrics": {
    "numBytesWritten": {
      "id": 3,
      "name": "number of bytes written",
      "value": 0
    },
    "numRecordsWritten": {
      "id": 4,
      "name": "number of records written",
      "value": 0
    }
  },
  "jdbcWrapper": "net.snowflake.spark.snowflake.DefaultJDBCWrapper$@2da464e2",
  "params": "Snowflake Data Source",
  "userSchema": "None",
  "sqlContext": "org.apache.spark.sql.SQLContext@415f672d",
  "log": "org.apache.logging.slf4j.Log4jLogger@10cc51c0",
  "pushdownStatement": "SELECT * FROM ( workorder_data_gold ) AS \"SF_CONNECTOR_QUERY_ALIAS\"",
  "bitmap": 7
}
```

The only data specific to the source table is that `pushdownStatement` which could be parsed to get the table name, but not enough for the datasource URI. With that in mind I decided to leave the READ capturing out of this PR.

Reply:
I don't know have an explicit answer to your question. However, I did some digging on capturing READ, and I wasn't able to find a way to capture the data source URI given the underlying `net.snowflake.spark.snowflake.SnowflakeRelation` that gets matched on when the logical plan includes reading from a snowflake source. 

This is what was available in that `SnowflakeRelation` object on a READ test I executed:

```
{
  "schema": {
    "fields": [
      {
        "name": "WORKORDERID",
        "type": "StringType",
        "nullable": true
      },
      {
        "name": "PRODUCTID",
        "type": "StringType",
        "nullable": true
      },
      {
        "name": "ORDERQTY",
        "type": "StringType",
        "nullable": true
      },
      {
        "name": "SCRAPPEDQTY",
        "type": "StringType",
        "nullable": true
      },
      {
        "name": "STARTDATE",
        "type": "StringType",
        "nullable": true
      },
      {
        "name": "ENDDATE",
        "type": "StringType",
        "nullable": true
      },
      {
        "name": "MODIFIEDDATE",
        "type": "StringType",
        "nullable": true
      },
      {
        "name": "BONUS_PERCENT",
        "type": "DoubleType",
        "nullable": true
      }
    ]
  },
  "scanMetrics": {
    "numBytesRead": {
      "id": 1,
      "name": "number of bytes read",
      "value": 7197975
    },
    "numRecordsRead": {
      "id": 2,
      "name": "number of records read",
      "value": 72591
    }
  },
  "writeMetrics": {
    "numBytesWritten": {
      "id": 3,
      "name": "number of bytes written",
      "value": 0
    },
    "numRecordsWritten": {
      "id": 4,
      "name": "number of records written",
      "value": 0
    }
  },
  "jdbcWrapper": "net.snowflake.spark.snowflake.DefaultJDBCWrapper$@2da464e2",
  "params": "Snowflake Data Source",
  "userSchema": "None",
  "sqlContext": "org.apache.spark.sql.SQLContext@415f672d",
  "log": "org.apache.logging.slf4j.Log4jLogger@10cc51c0",
  "pushdownStatement": "SELECT * FROM ( workorder_data_gold ) AS \"SF_CONNECTOR_QUERY_ALIAS\"",
  "bitmap": 7
}
```

The only data specific to the source table is that `pushdownStatement` which could be parsed to get the table name, but not enough for the datasource URI. With that in mind I decided to leave the READ capturing out of this PR.

Reply:
Hm, this is interesting:
> "jdbcWrapper": "net.snowflake.spark.snowflake.DefaultJDBCWrapper$@2da464e2"

Is Snowflake connection just a wrapper over JDBC? Can you introspect that wrapper object? I believe there will be IPs, connection string etc.

Basically, what would be ideal to achieve (though not always possible unfortunately depending on the technology) is that the data source URI that comes out from the WRITE operation to match exactly the URL that comes out from a respective READ operation (the one that reads the logically same thing that was written by the write op).
The structure on that URL doesn't really matter as long as it resembles a URL and reliably identify the data entity that is being read or written. That helps Spline to connect execution plans based on the data source URL match, and build the end-to-end lineage.